### DISCUSSION TITLE
One pool, multiple systems with nested datasets. How to configure mounting?

### DISCUSSION THREAD

**Discussion by LunarLambda** [openzfs/zfs#16866] [Q&A]
Suppose I wish to have two separate Linux installations on the same pool, under `tank/foo` and `tank/bar`. Further suppose I wish to have separate datasets on each system, such as `/home` or `/var/log` or `/var/lib/something`.

Naively, these would have the same mountpoint set on both systems. As I understand, this would not work as is.

What is the recommended way to approach this kind of setup? Use `mountpoint=legacy` and `fstab`? Use `zfs-mount-generator`?

Another user I asked suggested setting `canmount=noauto` and either writing a systemd unit that executes `/sbin/zfs mount <datasets>`, with dependencies set up such that it "acts like" a `systemd.mount` unit or to use `mountpoint=legacy` but use `systemd.mount` units instead of `fstab`.

Looking online you can find people talking about installing multiple systems on a single pool however they tend to simply have the entire system be on a single dataset.

**Comment by IvanVolosyuk** [comment]
You can have something like:
pool/system/gentoo/ROOT -> legacy (kernel param: root=ZFS=pool/system/debian/var-log)
pool/system/gentoo/ROOT/home -> inherit (will mount automatically)
pool/system/gentoo/var-log (legacy, mount via fstab to /var/log)
pool/system/gentoo/something -> (legacy, mount via fstab to /var/lib/something)
pool/opt -> (mountpoint: /opt, automatically mounted)
pool/system/debian ... same stuff

Things to consider:
* If entire system is in the same dataset, in case you use snapshots and want to rollback to a good known state, it is useful to have it all together in a single dataset as rollback will be an atomic operation. With multiple datasets the snapshots and rollbacks is not strictly atomic as far as I understand, this may cause potential discrepancies. 
* I would split home away from the system. When I rollback the system, I usually don't want to rollback my home.
* Also, if you want to boot an older version of system temporary, you will have to use 'zfs clone' and if it is not a single dataset with /home underneath - it will become way more difficult to do that. With single dataset for system, you can do something as simple as:
```zfs clone pool/system/gentoo/ROOT@2024-12-01 pool/system/gentoo/ROOT.old``` and change kernel root parameter to root=ZFS=pool/system/gentoo/ROOT.old (Boot and check that the problem went away at older system version)

Having /var/log, etc.. on a different mount point has some benefits: e.g you can have different recordsize and compression settings.
Have a look at zfsbootmenu, which supports multiple systems. If you plan to use it, you probably want to keep your systems as single datasets.

The nice part, you can experiment, try several variants, see which works for you. There is 'zpool checkpoint' feature, so if you want to do some drastic changes to your layout, you can checkpoint the pool before doing that and revert back to it if it doesn't work.
